src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp

```diff
@@ -0,0 +1,256 @@
+    }
+
+    // Get parent's hash
+    std::size_t seed = 0;
+    if (m_parent) {
+        seed = m_parent->get_hash();
+    } else {
+        seed = std::hash<void*>()(m_orig_data) + 0x9e3779b9;
+        seed ^= std::hash<std::string>()(m_orig_shape.to_string()) + 0x9e3779b9;
+        seed ^= std::hash<std::string>()(m_orig_type.to_string()) + 0x9e3779b9;
+    }
+
+    // Combine with this hash
+    seed ^= std::hash<int>()(static_cast<int>(m_transform.first)) + 0x9e3779b9;
+    if (m_transform.first == TransformType::PERMUTE) {
+        const auto& axes = std::get<std::vector<std::size_t>>(m_transform.second);
+        for (const auto& axis : axes) {
+            seed ^= std::hash<std::size_t>()(axis) + 0x9e3779b9;
+        }
+    } else if (m_transform.first == TransformType::CONCAT) {
+        const auto& axis = std::get<ConcatMeta>(m_transform.second).second;
+        seed ^= std::hash<std::size_t>()(axis) + 0x9e3779b9;
+        for (auto& lt : std::get<ConcatMeta>(m_transform.second).first) {
+            seed ^= lt.get_hash() + 0x9e3779b9;
+        }
+    }
+
+    return m_hash;
+}
+}  // namespace weights
+}  // namespace npuw
+}  // namespace ov
+
+using ov::npuw::weights::LazyTensorImpl;
+
+LazyTensorImpl::LazyTensorImpl(const TransformType& type, const Transform& transform) {
+    if (type == TransformType::TENSOR && std::holds_alternative<LTData>(transform)) {
+        m_transform = std::make_pair(type, transform);
+        ov::Tensor tensor;
+        if (std::holds_alternative<ConstPtr>(std::get<LTData>(transform))) {
+            tensor = ov::npuw::util::tensor_from_const(std::get<ConstPtr>(std::get<LTData>(transform)));
+        } else {
+            tensor = std::get<ov::Tensor>(std::get<LTData>(transform));
+        }
+        m_orig_data = tensor.data();
+        m_orig_shape = tensor.get_shape();
+        m_orig_type = tensor.get_element_type();
+    } else if (type == TransformType::CONCAT && std::holds_alternative<ConcatMeta>(transform)) {
+        m_transform = std::make_pair(type, transform);
+    } else {
+        NPUW_ASSERT(false);
+    }
+
+    m_hash = get_hash();
+}
+
+bool LazyTensorImpl::operator==(const LazyTensorImpl& other) const {
+    if (m_hash != other.m_hash || m_orig_data != other.m_orig_data || m_orig_shape != other.m_orig_shape ||
+        m_orig_type != other.m_orig_type || m_transform.first != other.m_transform.first) {
+        return false;
+    }
+
+    ConcatMeta m1, m2;
+
+    switch (m_transform.first) {
+    case TransformType::TENSOR:
+        // everything is already compared above - skip
+        break;
+    case TransformType::CONVERT:
+        // everything is already compared above - skip
+        break;
+    case TransformType::PERMUTE:
+        if (std::get<std::vector<std::size_t>>(m_transform.second) !=
+            std::get<std::vector<std::size_t>>(other.m_transform.second)) {
+            return false;
+        }
+        break;
+    case TransformType::CONCAT:
```

Comment:
you certainly underused variants which were proposed initially.

a transformation on the tensor is a variant.

Reply:
Are you suggesting making transformation a proper class?